### ISSUE TITLE
libyang2 fails to load modules

### ISSUE THREAD

**Issue by choppsv1** [CESNET/libyang#1528]
With the following commit 1ccbf54

my FRR unit tests are failing in the following way:

2021/04/22 10:07:35 libyang: Invalid argument parent || module || ext (lys_getnext_()).
2021/04/22 10:07:35 libyang: Not found node "lib" in path. (Schema location /frr-routing:routing/control-plane-protocols/control-plane-protocol/frr-bgp:bgp/global/afi-safis/afi-safi/ipv4-flowspec/flow-spec-config/interface.)
2021/04/22 10:07:35 [Y3P23-A91R3][EC 100663318] yang_module_load: failed to load data model: frr-bgp

In order to reproduce one can run FRR unit test:

```
git clone https://github.com/LabNConsulting/frr.git
cd frr
git checkout chopps/ly2-triage
autoreconf -i
mkdir build
cd build
CPPFLAGS="$(pkg-config --cflags-only-I libyang)" ../configure --prefix=/usr --enable-dev-build --localstatedir=/var/run/frr --sbindir=/usr/lib/frr --sysconfdir=/etc/frr --enable-vtysh --enable-pimd --enable-sharpd --enable-snmp=agentx --enable-multipath=64 --enable-user=frr --enable-group=frr --enable-vty-group=frrvty --with-pkg-extra-version=-my-manual-build
make -j16
make check #<-- it will run and things will fail, then the specific test to run is:
tests/bgpd/.libs/test_peer_attr
```

I added an abort to the CHECK_RET macro so it would stop at the failure point and then I turned all debugging on in libyang and reran the test. I've attached that large debug output.

[ly2-debug.txt](https://github.com/CESNET/libyang/files/6391865/ly2-debug.txt)

**Comment by choppsv1** [comment]
If you would also like to enable debug in the test program here is the diff to do that:
```
modified   tests/bgpd/test_peer_attr.c
@@ -18,6 +18,7 @@
  */
 #include <zebra.h>
 
+#include "debug.h"
 #include "memory.h"
 #include "plist.h"
 #include "bgpd/bgpd.h"
@@ -1459,6 +1460,8 @@ int main(void)
 	struct test_peer_attr *pa, *pac;
 	struct listnode *node, *nnode;
 
+	DEBUG_ON(&nb_dbg_events);
+	yang_debugging_set(1);
 	bgp_startup();
 
 	pa_list = list_new();
```

**Comment by michalvasko** [comment]
Thanks for the effort but simply loading the modules with `yanglint` was enough to reproduce. Should now be fixed with a test included. Also, `lyd_value_get_canonical()` was being called with `NULL` value that was crashing but now should log invalid argument and return `NULL`, you probably want to fix that.

**Comment by choppsv1** [comment]
FRR is adding data to the lysc_node->priv field, I'm seeing a crash due to that data being NULL, is it possible for this new recompile thing to be replacing previous compiled nodes with new ones which do not have this private data set anymore?

**Comment by michalvasko** [comment]
Yes, if you create some context, set `priv` pointers in it, and then change the context again, it may be recompiled and all the `priv` pointers lost.

**Comment by choppsv1** [comment]
How do I know if this is happening? I'll admit I have no idea what this recompile change is about. :)

**Comment by choppsv1** [comment]
I guess what I mean is what constitutes "changing the context"?

**Comment by michalvasko** [comment]
It is not only about this change, it was there even before but it was changing a bit. Currently the rule should be only one, `if there is a module being compiled (implemented) that imports an already-compiled module, the whole context is recompiled`.

For future, it should be possible to optimize so that not the whole context is recompiled but that will require caution because as you saw, it is too easy to get it wrong. And it should affect only the recompilation time, nothing else.

**Comment by choppsv1** [comment]
So to clarify, does this mean if I load a module at some point, the side-effect might be that previously loaded modules get recompiled? If so I think I need a callback when this happens so that I can redo the priv data on those modules I've already done this for.

**Comment by michalvasko** [comment]
Hm, that is a possibility but the general rule should be to set `priv` pointers only after you are done changing the context and if that is not possible, always set them for the whole context if you have changed it and need those `priv` pointers set again.

**Comment by choppsv1** [comment]
I guess my point here is "changing the context" might be a lot more obvious to you than it is to me. :) How do I know if I've "changed the context". A callback leaves no doubt.

**Comment by choppsv1** [comment]
I believe this is happening here, after `frr-isis` is loaded I see a recompile, frr-interface was loaded previous to this. The sigsegv is on a schema node present in frr-interface.

```
Breakpoint 2, yang_module_load (module_name=0x7ffff7b27ca5 "frr-interface") at ../lib/yang.c:116
116                     ly_ctx_load_module(ly_native_ctx, module_name, NULL, NULL);
(gdb) c
Continuing.

Breakpoint 2, yang_module_load (module_name=0x5555556207c8 "frr-isisd") at ../lib/yang.c:116
116                     ly_ctx_load_module(ly_native_ctx, module_name, NULL, NULL);
(gdb) c
Continuing.

Breakpoint 1, lys_recompile (ctx=0x5555558c2fc0, log=1 '\001') at /home/chopps/w/ly2/src/schema_compile.c:1499
1499    {
(gdb) c
Continuing.

Breakpoint 2, yang_module_load (module_name=0x7ffff7b4026e "frr-route-map") at ../lib/yang.c:116
116                     ly_ctx_load_module(ly_native_ctx, module_name, NULL, NULL);
(gdb) c
Continuing.

Breakpoint 2, yang_module_load (module_name=0x7ffff7b44e32 "frr-vrf") at ../lib/yang.c:116
116                     ly_ctx_load_module(ly_native_ctx, module_name, NULL, NULL);
(gdb) c
Continuing.

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7aa9b25 in nb_candidate_edit (candidate=0x5555559006e0, nb_node=0x0, operation=NB_OP_CREATE,
    xpath=0x7fffffff6e10 "/frr-interface:lib/interface[name='r1-eth5'][vrf='default']/frr-isisd:isis", previous=0x0, data=0x555555acd2d0)
    at ../lib/northbound.c:686
```

**Comment by michalvasko** [comment]
Well, to put it as simply as I can, a recompile may happen after calling `lys_parse*()` or `ly_ctx_load_module()` functions (adding a module) or `lys_set_implemented()` (implementing a new module or changing its enabled feature set). That should be all, unless I have forgotten about some. I suppose a comment could be added to these functions to warn about the recompilation.

**Comment by choppsv1** [comment]
We discussed this in the FRR YANG meeting today, and we really think a callback whenever a module is [re]compiled makes imminent sense here. We can fixup the priv fields whenever they need to be, then and don't need to follow any indirect logic (or future changing reasons for recompilation etc).

**Comment by michalvasko** [comment]
I am not sure this is really the best solution. Recompilation can never be caused by anything else or other functions than those I mentioned. What we were changing was only the exact cases when it happens. If you expect in the code it happens always after calling the functions, you will be safe.

Also, to get faster context creation times, you should be using the context flag `LY_CTX_EXPLICIT_COMPILE`. Then, after you are done "changing the context" (calling the aforementioned functions), you call `ly_ctx_compile()` and you can set the `priv` pointers safely afterwards. This is the best way of creating the context.

**Comment by choppsv1** [comment]
> I am not sure this is really the best solution. Recompilation can never be caused by anything else or other functions than those I mentioned. What we were changing was only the exact cases when it happens. If you expect in the code it happens always after calling the functions, you will be safe.

Even when you mentioned it you said "unless I have forgotten about some". This isn't a solid foundation for APIs. Its requiring too much inner knowledge of the implementation on the users part, especially when the author of the implementation would need to go double check code to make sure.

Also, am I supposed to redo all of my schema private data every time I load a module? This is not required unless that load caused a recompile (they all don't), but because I don't know if the loading of the module has had a recompile side-effect I have to do it anyway.

Instead, libyang could just tell me when it has recompiled a module, and I can do the private data association then. This seems really clean to me, not sure why you'd object.

**Comment by michalvasko** [comment]
> Even when you mentioned it you said "unless I have forgotten about some". This isn't a solid foundation for APIs. Its requiring too much inner knowledge of the implementation on the users part, especially when the author of the implementation would need to go double check code to make sure.

I answered to my best knowledge and truthfully. Have you agreed with the solution I proposed, I would have looked through the code to find all the functions that could cause the recompilation and add a comment warning about it.

> Also, am I supposed to redo all of my schema private data every time I load a module? This is not required unless that load caused a recompile (they all don't), but because I don't know if the loading of the module has had a recompile side-effect I have to do it anyway.

No, what I proposed was to strictly differentiate between creating (and modifying) the context and working with it (and any `priv` pointers in the modules). We will see in future but it is possible the best solution will turn out to be to forbid modifying a compiled context and allow only to always free all the compiled modules and recompiled it whole, in which case you will no longer be given any choice and will have to set `priv` pointer every time.

> Instead, libyang could just tell me when it has compiled a module, and I can do the private data association then. This seems really clean to me, not sure why you'd object.

I objected because this callback supports using the context in using it not the intended way. However, since there are most likely still some problems even in the API preventing from working with the context the way I proposed for every use-case, I will add the callback. I am going to assume you cannot use the proposed solution for your specific use-case.
